Support UK radar data

BernardClaxton opened the pull request:
Fixes #1759

### Contribution checklist

Aim to have all relevant checks ticked off before merging. See the [developer's guide](https://metoffice.github.io/CSET/contributing/) for more detail.

- [ ] Documentation has been updated to reflect change.
- [ ] New code has tests, and affected old tests have been updated.
- [ ] All tests and CI checks pass.
- [ ] Ensured the pull request title is descriptive.
- [ ] Conda lock files have been updated if dependencies have changed.
- [ ] Attributed any Generative AI, such as GitHub Copilot, used in this PR.
- [ ] Marked the PR as ready to review.

BernardClaxton committed:
add Nimrod radar page to CSET GUI #1759

BernardClaxton committed:
put nimrod functions in fetch_data.py to grab radar files #1759

BernardClaxton committed:
added job in cylc workflow to grab nimrod obs #1759

BernardClaxton committed:
defined fetchnimrod task as fetch-nimrod.py #1759

BernardClaxton committed:
put in script fetch-nimrod.py #1759

BernardClaxton committed:
edits to fetch_data.py #1759

BernardClaxton committed:
edits to fetch-nimrod.py #1759

BernardClaxton committed:
edits to flow-cylc #1759

BernardClaxton committed:
edits to meta/rose-meta.conf #1759

BernardClaxton committed:
fetch-nimrod.py update #1759

BernardClaxton committed:
constants.py added #1759

BernardClaxton committed:
added nimrod colorbar #1759

BernardClaxton committed:
added rose-meta colorbar #1759

BernardClaxton committed:
adding cube_utils

BernardClaxton committed:
recipe for nimrod #1759

BernardClaxton committed:
debugging code added to fetch-nimrod.py at L715 #1759

BernardClaxton committed:
debugging code added to fetch-nimrod.py at L715 #1759

BernardClaxton committed:
Merge remote-tracking branch 'origin/main' into 1759_import_nimrod

Sylviabohnenstengel commented:
for info if the upgrade of you branch with main causes problems then I have created a branch of your branch and merged main in now including the addition of plotting transverse mercator. So you should be able to plot the radar data. 
Branch: #1759_import_nimrod_merge_main

jfrost-mo reviewed (commented):
`src/CSET/cset_workflow/app/fetch_fcst/bin/fetch_data.py` is the main file for this change, as the stuff in fetch_nimrod is largely unused.

What is our plan with this? Are we going to try and reuse the RES capability in cube_utils? Or do we not really need it given the existing capabilities of CSET?

The recipe and such looks broadly sensible, though I think we can get away with hard-coding some more for the initial simple recipe.

BernardClaxton committed:
Nimrod fetching code moved into fetch_nimrod.py #1759

BernardClaxton committed:
tidied the Nimrod radio buttons in GUI #1759

BernardClaxton committed:
rewrote fetch_nimrod.py to output Nimrod data as NetCDF #1759

BernardClaxton committed:
rewrote Nimrod stuff in flow.cylc #1759

BernardClaxton committed:
put simple Nimrod loaders in spatial_field.py #1759

BernardClaxton committed:
modified plot.py to produce Nimrod radar histograms #1759

Sylviabohnenstengel committed:
suggestion how to include any for two variables in teh plot function to enable radar plotting.

BernardClaxton committed:
popped in fix to cope with both model and radar rainfall  #1759

BernardClaxton committed:
popped Nimrod line plots into histograms.py #1759

BernardClaxton committed:
popped Nimrod 2-D plots into loader spatial_field.py #1759

BernardClaxton committed:
adding new recipe for 2-D Nimrod rainfall plots #1759

BernardClaxton committed:
updating fetch script for 2-D Nimrod radar data #1759

BernardClaxton committed:
Nimrod code in histograms loader #1759

BernardClaxton committed:
Nimrod code in spatial_field loader #1759

BernardClaxton committed:
Nimrod recipe for plotting spatial field of hourly rainfall #1759

BernardClaxton committed:
Nimrod entries in Rose GUI #1759

BernardClaxton committed:
adding the radar loader file .../loaders/radar.py  #1759

BernardClaxton committed:
removing the radar recipe from the histogram loader file  #1759

BernardClaxton committed:
removing the radar recipe from the spatial_field loader file  #1759

BernardClaxton committed:
adding the radar loader file  ..../loaders/radar.py #1759

BernardClaxton committed:
tidied the labelling of the colour bar for model rainfall spatial plots  #1759

jfrost-mo committed:
Remove unused cube_utils.py file

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/cube_utils.py:
Removed in 3e3a6d76f57750241d20eebcfbf7875ecf6df781.

jfrost-mo committed:
Merge branch 'main' into 1759_import_nimrod

The radar configuration was moved to the Observation rose meta
directory.

jfrost-mo committed:
Initialise namespace through unpacking to support python 3.12

Using a positional argument was only added in python 3.13.

jfrost-mo committed:
Update namespace of NIMROD radar obs

This matches the new location of the other observation settings.

jfrost-mo commented on src/CSET/cset_workflow/meta/observations/rose-meta.conf:
```diff
@@ -86,6 +86,53 @@ compulsory=true
+help=Switch to select default gridded UK Nimrod rainfall accumulation.
+compulsory=true
+values="rainaccum_comp_hour", ""
+value-titles=rainaccum_comp_hour, off
```
Should this be a boolean toggle? Or are there cases when we will have multiple options?

BernardClaxton replied:
<img width="765" height="421" alt="image" src="https://github.com/user-attachments/assets/f21cfdd9-cb08-451b-a9aa-4c41d5e92a3e" />


This has been done - the Nimrod sources are now selected in the Rose GUI using boolean toggles.

jfrost-mo commented on src/CSET/operators/plot.py:
```diff
@@ -1397,6 +1403,8 @@ def _plot_and_save_histogram_series(
     model_colors_map = _get_model_colors_map(cubes)
 
+    print("Using internal function _plot_and_save_histogram_series")
+
```
```suggestion
```
Remove print statements.

jfrost-mo commented on src/CSET/operators/plot.py:
```diff
@@ -1405,7 +1413,8 @@ def _plot_and_save_histogram_series(
         # than seeing if long names exist etc.
         # Exception case, where distribution better fits log scales/bins.
-        if "surface_microphysical" in title:
+        print("hello - title is ", title)
```
```suggestion
```

jfrost-mo committed:
Combine comprehensions

jfrost-mo commented on src/CSET/recipes/surface_fields/generic_surface_spatial_plot_sequence_radar_rainfall.yaml:
```diff
@@ -0,0 +1,34 @@
+#      level_constraint:
+#        operator: constraints.generate_level_constraint
+#        coordinate: "pressure"
+#        levels: []
```
```suggestion
```

jfrost-mo commented on src/CSET/recipes/surface_fields/generic_surface_spatial_plot_sequence_radar_rainfall.yaml:
```diff
@@ -0,0 +1,34 @@
+category: Surface Spatial Plot
+#title: $MODEL_NAME $VARNAME $METHOD
+title: $RADAR_NAME $VARNAME
```
```suggestion
title: $RADAR_NAME $VARNAME
```
This title could be more descriptive.

jfrost-mo commented on src/CSET/recipes/surface_fields/generic_surface_spatial_plot_sequence_radar_rainfall.yaml:
```diff
@@ -0,0 +1,34 @@
+
+#  - operator: collapse.collapse
+#    coordinate: [time]
+#    method: $METHOD
```
```suggestion
```
We could try and make this generic over collapse method. Now #2027 is in we should be more reliable for handling various input data.

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,104 @@
+    #
+    # The different sources of Nimrod rainfall acculumulation have
+    # different spatial grids. So each source requires its own yaml
+    # realisation to prevent incompatible cubes being created.
```
```suggestion
    # different spatial grids. So each source requires its own
    # recipe to prevent incompatible cubes being created.
```

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,104 @@
+    # Histograms for surface (2D) Nimrod radar rainfall.
+    #
+    # To get multiple radar sources plotted on the histogram the yaml
+    # realisation must be done by passing lists of the radar_ids and
```
```suggestion
    # To get multiple radar sources plotted on the histogram the
    # recipe must be done by passing lists of the radar_ids and
```

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,104 @@
+from CSET.recipes import Config, RawRecipe
+
+
+def get_radar_sources(conf) -> dict:
```
Given we want the name, id, and varname for each radar together, we can pivot how we are creating the data structure to make it easier to use. We can also use the trick that `bool([]) == False` to avoid needing to explicitly set radar_flag.

This can then replace this data structure with a simple list of dicts. E.g.

```py
radar_sources = []

if conf.NIMROD_COMP_1KM:
    radar_sources.append({
        "name": "Nimrod_1km",
        "id": "Nimrod1km",
        "varname": "Hourly rain accumulation"
    })

...
```

Then when you use it have each radar's information together.

```py
for radar in radar_sources:
    # Do something with the name, id, and varname.
    print(f"name = {radar['name']}")
    ...
```

jfrost-mo replied:
Done in 785ffc77d582403b698aa7dac6ac75ec68b4118b.

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+    )
+    # cube_obs_weighted.data = np.ma.masked_where(
+    #                         weights < weight_min,
+    #                         cube_obs_weighted.data)
```
```suggestion
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+            # Process Nimrod data between the start and end dates.
+            date_use = date_start
+            while date_use <= date_end:
+                # print("date is ", date_use)
```
```suggestion
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+        "date_type": "initiation",
+        "data_time": datetime.fromisoformat(os.environ["CYLC_TASK_CYCLE_POINT"]),
+        "forecast_length": isodate.parse_duration(os.environ["ANALYSIS_LENGTH"]),
+        #        "model_identifier": "Nimrod",
```
```suggestion
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+    # Note: if the weights are packed, then the maximum value
+    #       found will be 13 / 32 = 0.40625, i.e. all values < 1.
+    weights = cube_wei.data
+    if np.array(cube_wei.data).max() < 1.0:
```
```suggestion
    if weights.max() < 1.0:
```
The cube.data is already an array.

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+    #       found will be 13 / 32 = 0.40625, i.e. all values < 1.
+    weights = cube_wei.data
+    if np.array(cube_wei.data).max() < 1.0:
+        weights = (cube_wei.data * 32).round().astype(int)
```
```suggestion
        weights = (weights * 32).round().astype(int)
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+                filename_obs_nc = Path(filename_obs_nc).with_suffix(".nc")
+                iris.save(nimrod_cube_obs, filename_obs_nc)
+
+    return
```
```suggestion
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+                    + nimrod_dict[nimrod_field]["obs_dir"]
+                    + "/"
+                    + f"{date_use.year}/{date_use.strftime('%Y%m%d%H%M')}"
+                    + nimrod_dict[nimrod_field]["obs_fname"]
```
```suggestion
                    f"{nimrod_dict[nimrod_field]["basedir"]}/"
                    f"{nimrod_dict[nimrod_field]["obs_dir"]}/"
                    f"{date_use.year}/{date_use.strftime('%Y%m%d%H%M')}"
                    f{nimrod_dict[nimrod_field]["obs_fname"]}"
```
Rather than concatenate at runtime we can use a string literal broken across multiple lines.

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+                        + nimrod_dict[nimrod_field]["weights_dir"]
+                        + "/"
+                        + f"{date_use.year}/{date_use.strftime('%Y%m%d%H%M')}"
+                        + nimrod_dict[nimrod_field]["weights_fname"]
```
```suggestion
                        f"{nimrod_dict[nimrod_field]["basedir"]}/"
                        f"nimrod_dict[nimrod_field]["weights_dir"]}/"
                        f"{date_use.year}/{date_use.strftime('%Y%m%d%H%M')}"
                        f"{nimrod_dict[nimrod_field]["weights_fname"]}"
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+                        + f"{date_use.strftime('%Y%m%d%H%M')}"
+                        + "_"
+                        + nimrod_dict[nimrod_field]["obs_id"]
+                        + "_weights"
```
```suggestion
                        f"{nimrod_dir_wei}/{date_use.strftime('%Y%m%d%H%M')}"
                        f"_{nimrod_dict[nimrod_field]["obs_id"]}_weights"
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+                    + "/"
+                    + f"{date_use.strftime('%Y%m%d%H%M')}"
+                    + "_"
+                    + nimrod_dict[nimrod_field]["obs_id"]
```
```suggestion
                    f"{nimrod_dir}/{date_use.strftime('%Y%m%d%H%M')}"
                    f"_{nimrod_dict[nimrod_field]["obs_id"]}"
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+    #* DATA_PERIOD
+    #* DATE_TYPE
+    #* MODEL_IDENTIFIER
+    * ROSE_DATAC
```
Check this list is up-to-date.

BernardClaxton replied:
This list has now been updated.

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+    for nimrod_field in v["field"]:
+        if nimrod_field:
+            logging.debug("Processing Nimrod field: %s", nimrod_field)
+            print("processing Nimrod field ", nimrod_field)
```
```suggestion
```

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+                        + "_weights"
+                    )
+                    filename_wei_nc = Path(filename_wei_nc).with_suffix(".nc")
+                    iris.save(nimrod_cube_weights, filename_wei_nc)
```
Do we need to save the weights if we have already applied them to the data?

BernardClaxton replied:
After discussion with Sylvia we decided to retain saving the weights. In future it might be useful to plot the weights field to check the integrity of the radar observational data.

BernardClaxton replied:
I've now implemented plotting of the Nimrod weight files. There is a switch in the Rose GUI for selecting this operation, so it can be switched off. Here is an example:

<img width="1023" height="709" alt="image" src="https://github.com/user-attachments/assets/f14a3fbd-8ccc-4bc8-a5ab-a9c5b7255931" />

BernardClaxton replied:
Removed the (mm) label from the weights plot:

<img width="501" height="133" alt="image" src="https://github.com/user-attachments/assets/5d42b839-bc69-4796-a162-f6c172ed550b" />

jfrost-mo commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
```diff
@@ -0,0 +1,258 @@
+    return variables
+
+
+def _nimrod_dictionary() -> dict:
```
Should this read from a data file that can be put into the restricted repository as a site-specific file?

jfrost-mo reviewed (changes requested):
Looking sensible. I've left quite a few comments, but most of them are mechanical cleanup; I'll apply some myself to get the comment count down.

It would be good to see some tests for fetch_nimrod; we can work on them together to validate it is doing the correct things.

We also need to have a think about where fetch_nimrod will live. Its is a reasonable size piece of functionality, which makes me want to have it in the public repository, both for better integration into tests, as well as to allow it to serve as an example for others adding their own radar. On the other hand, it contains a number of Met Office specific tasks, and more broadly I think NIMROD is a Met Office specific format, which will limit the ability for it to be run elsewhere. My suggestion is to factor out the paths and such (which are mostly in the nimrod_dict already) into a separate data file and have that in the restricted repository, but leave the bulk of the code here.

jfrost-mo committed:
Remove commented out/useless/print code

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

jfrost-mo committed:
Avoid unnecessary array creation

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

jfrost-mo committed:
Reformat string templates

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

jfrost-mo committed:
Update terminology in comments

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

jfrost-mo committed:
Remove prints

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

jfrost-mo committed:
Clean up radar recipe files

jfrost-mo committed:
Create directory for weights when needed

This is mostly to satisfy my type checker that nimrod_dir_wei is set
when used.

jfrost-mo committed:
Merge branch 'main' into 1759_import_nimrod

jfrost-mo commented on src/CSET/cset_workflow/meta/observations/rose-meta.conf:
```diff
@@ -86,6 +86,53 @@ compulsory=true
+compulsory=true
+values="rainaccum_comp_hour_2km", ""
+value-titles=rainaccum_comp_hour_2km, off
+sort-key=nimrod-radar-5
```
We are missing the settings for `NIMROD_COMP_5MIN` from here.

jfrost-mo replied:
Added myself.

jfrost-mo committed:
Add NIMROD normalisation callback

The number of radar sites can change, preventing the cubes from merging.
Therefore the respective cube attributes are stripped.

jfrost-mo committed:
Add configuration for 5 minute rainfall rate

jfrost-mo committed:
Rearrange loader to use simpler structure

jfrost-mo committed:
Add 5 minute rainfall rate to loader

jfrost-mo committed:
Remove troublesome attributes from 5 minute rainrate cubes

jfrost-mo commented:
Be warned that plotting spatial plots for 5 minutely data takes ages.

jfrost-mo commented:
`Rainfall rate Composite` will also want a colourbar, so the 5 minutely data displays sensibly.

My run: https://wwwspice/~james.frost/CSET-radar-test/

BernardClaxton committed:
changed Composite to composite in name of Nimrod_5min variable #1759

BernardClaxton committed:
Added entry in _colorbar_definition.json for Nimrod_5min data #1759

BernardClaxton committed:
Added log rainfall colorbar for Nimrod_5min 2-D field plots #1759

BernardClaxton commented:
<img width="1023" height="701" alt="image" src="https://github.com/user-attachments/assets/509d5e3a-6535-4097-b5ca-a08459743801" />

Nimrod 5 minute rainfall rates now plots using the same logarithmic colourbar as the other rainfall plots.

BernardClaxton committed:
Changed the varname for Nimrod 5min data back to Rainfall rate Composite  #1759

BernardClaxton committed:
Changed the Nimrod Rose GUI buttons to boolean  #1759

BernardClaxton committed:
Tidied logic handling of the Nimrod toggles  #1759

BernardClaxton committed:
Removed commented out Nimrod lines in rose-meta.conf  #1759

BernardClaxton committed:
Updated comments for retrieve_nimrod  #1759

BernardClaxton committed:
Delete src/CSET/cset_workflow/app/fetch_nimrod/bin/constants.py

None of the three constants defined in this file are used. The defined constants are:

HOUR_IN_SECONDS = 3600
M_IN_KM = 1000
REAL_MATCH_TOL = 1.0e-4

BernardClaxton commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/constants.py:
```diff
@@ -0,0 +1,9 @@
+#: Number of seconds in an hour.
+HOUR_IN_SECONDS = 3600
+#: Number of metres in a kilometre.
+M_IN_KM = 1000
```
None of these three constants are used anywhere. The file constants.py has been deleted.

BernardClaxton committed:
Implemented plotting of Nimrod accumulation weight fields  #1759

BernardClaxton committed:
Changed an except from Exception to FileNotFoundError  #1759

BernardClaxton committed:
Refactored fetch_nimrod.py #1759

BernardClaxton committed:
Corrected the time stepping in fetch_nimrod.py #1759

BernardClaxton committed:
Change the units for Nimrod weights to dimensionless #1759

BernardClaxton committed:
moved the Met Office Nimrod paths data into a restricted file  #1759

BernardClaxton committed:
sensitive Nimrod handling data is now in a json file  #1759

BernardClaxton committed:
Added if __name__ == __main__: to fetch_nimrod.py  #1759

BernardClaxton committed:
Put Nimrod fetching tests in tests/test_fetch_nimrod.py   #1759

BernardClaxton committed:
Small edit to a comment in tests/test_fetch_nimrod.py   #1759

BernardClaxton committed:
Another NImrod retrieval test added #1759

BernardClaxton commented on src/CSET/cset_workflow/app/fetch_nimrod/bin/fetch_nimrod.py:
This has now been done. The Met Office specific data related to Nimrod has been put in a json file in the cset-restricted-files repository.

BernardClaxton committed:
Used pragma to exclude coverage reporting for restricted file #1759

BernardClaxton committed:
Add in timeseries plots of domain mean of Nimrod data  #1759

BernardClaxton committed:
Plot histograms of domain Nimrod 5 minute rainfall rate #1759

BernardClaxton commented:
Output from an example workflow can be found here https://wwwspice/~bernie.claxton/CSET_nimrod3/

BernardClaxton committed:
Merge branch 'main' into 1759_import_nimrod

daflack committed:
Merge branch 'main' into 1759_import_nimrod

jfrost-mo commented on tests/test_fetch_nimrod.py:
```diff
@@ -0,0 +1,164 @@
+# Tests for apply_radar_weights
+def make_cube(data):
+    """Form a cube of test data from an array."""
+    return SimpleNamespace(data=np.array(data))
```
Rather than mocking, we can use real Cubes throughout these tests.

```py
from iris.cube import Cube

obs = Cube([[1.0, 2.0], [3.0, 4.0]])
weights = Cube([[11, 12], [10, 13]])
```

BernardClaxton replied:
Real Cubes have been put in ```tests/test_fetch_nimrod.py```.

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,189 @@
+                    "SUBAREA_EXTENT": conf.SUBAREA_EXTENT
+                    if conf.SELECT_SUBAREA
+                    else None,
+                },
```
```suggestion
                    "SUBAREA_NAME": conf.SUBAREA_NAME if conf.SELECT_SUBAREA else "",
                },
```
We should also provide the `SUBAREA_NAME` (Added since this PR was last updated).

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,189 @@
+                else "realization",
+                "SUBAREA_TYPE": conf.SUBAREA_TYPE if conf.SELECT_SUBAREA else None,
+                "SUBAREA_EXTENT": conf.SUBAREA_EXTENT if conf.SELECT_SUBAREA else None,
+            },
```
```suggestion
                "SUBAREA_NAME": conf.SUBAREA_NAME if conf.SELECT_SUBAREA else "",
            },
```
We should also provide the `SUBAREA_NAME` (Added since this PR was last updated).

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,189 @@
+                else "realization",
+                "SUBAREA_TYPE": conf.SUBAREA_TYPE if conf.SELECT_SUBAREA else None,
+                "SUBAREA_EXTENT": conf.SUBAREA_EXTENT if conf.SELECT_SUBAREA else None,
+            },
```
```suggestion
                "SUBAREA_NAME": conf.SUBAREA_NAME if conf.SELECT_SUBAREA else "",
            },
```
We should also provide the `SUBAREA_NAME` (Added since this PR was last updated).

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,189 @@
+                "MODEL_NAME": [radar["name"] for radar in accum_radars],
+                "SUBAREA_TYPE": conf.SUBAREA_TYPE if conf.SELECT_SUBAREA else None,
+                "SUBAREA_EXTENT": conf.SUBAREA_EXTENT if conf.SELECT_SUBAREA else None,
+            },
```
```suggestion
                "SUBAREA_NAME": conf.SUBAREA_NAME if conf.SELECT_SUBAREA else "",
            },
```
We should also provide the `SUBAREA_NAME` (Added since this PR was last updated).

jfrost-mo commented on src/CSET/loaders/radar.py:
```diff
@@ -0,0 +1,189 @@
+                "MODEL_NAME": "Nimrod_5min",
+                "SUBAREA_TYPE": conf.SUBAREA_TYPE if conf.SELECT_SUBAREA else None,
+                "SUBAREA_EXTENT": conf.SUBAREA_EXTENT if conf.SELECT_SUBAREA else None,
+            },
```
```suggestion
                "SUBAREA_NAME": conf.SUBAREA_NAME if conf.SELECT_SUBAREA else "",
            },
```
We should also provide the `SUBAREA_NAME` (Added since this PR was last updated).

jfrost-mo commented on src/CSET/recipes/surface_fields/radar_mean_time_series.yaml:
```diff
@@ -0,0 +1,34 @@
+category: Surface Time Series
+title: Domain mean $VARNAME time series
```
```suggestion
title: Domain mean $VARNAME time series $SUBAREA_NAME
```
Include the subarea name in the title if present.

jfrost-mo commented on src/CSET/recipes/surface_fields/generic_surface_spatial_plot_sequence_radar_rainfall.yaml:
```diff
@@ -0,0 +1,20 @@
+category: Surface Spatial Plot
+title: $RADAR_NAME $VARNAME
```
```suggestion
title: $RADAR_NAME $VARNAME $SUBAREA_NAME
```
Include the subarea name in the title if set.

jfrost-mo reviewed (commented):
Looking nearly ready to merge. The pull request will need updating with the latest version of main, and then `SUBAREA_NAME` will want adding to the loaders and recipes.

The tests can also be simplified slightly by not mocking Cubes.

BernardClaxton committed:
Put real Cubes in test_fetch_nimrod.py  #1759

BernardClaxton committed:
Tidied test_fetch_nimrod.py  #1759

BernardClaxton committed:
Update src/CSET/loaders/radar.py

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

BernardClaxton committed:
Update src/CSET/loaders/radar.py

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

BernardClaxton committed:
Update src/CSET/loaders/radar.py

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

BernardClaxton committed:
Update src/CSET/loaders/radar.py

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

BernardClaxton committed:
Update src/CSET/loaders/radar.py

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

BernardClaxton committed:
Update src/CSET/recipes/surface_fields/radar_mean_time_series.yaml

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

BernardClaxton committed:
Update src/CSET/recipes/surface_fields/generic_surface_spatial_plot_sequence_radar_rainfall.yaml

Co-authored-by: James Frost <james.frost@metoffice.gov.uk>

jfrost-mo reviewed (commented):
Looking good. I'll bring this branch up-to-date and fix the conflicts, then approve it.

jfrost-mo committed:
Merge branch 'main' into 1759_import_nimrod

jfrost-mo reviewed (approved):
Code changes looking good, but you'll want to make sure rose-suite.conf.example is up-to-date.

See the error at https://github.com/MetOffice/CSET/actions/runs/29494779877/job/87609050167?pr=1825

You can run these checks locally with `make test-workflow`.

BernardClaxton committed:
Updated rose-suite.conf.example  #1759
